### PR TITLE
Refactor expr system

### PR DESCRIPTION
**Pull request by zhexuany**
### What problem does this PR solve? 
This PR makes current expression system being aware of data type info. By doing this, we can take advantage of spark's type coercer system. 

After this PR, `ExpressionTypeCoercer` is not used any more and can be safely deleted. `MetaResolver` will be only called in `ExprUtils`. 

We also convert `Expression` from `interface` to  an `abstract class`. In addition, `dataType` and `resolved` are added as filed of `Expression`. 

Some changes need to be noted:
1. `hasCode` and `equals` of `ColumnRef`. When a `ColumnRef` is resolved, we use its `name` and `dataType` to do the calculation. If it is not resolved, we only use its `name` to do the calculation. 
2. In `TiDAGRequest`, we use the name of `ColumnRef` rather than `ColumnRef` itself in Map, it is because if there are to identical name or variation, Spark's optimizer will throw `AnalysisException` before reaching the logic of `TiDAGRequest`.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/BasicExpression.scala [hunk 1]
```diff
@@ -1,31 +1,17 @@
...
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-package org.apache.spark.sql.tispark
+package org.apache.spark.sql.catalyst.expressions
```
Comment (birdstorm) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/BasicExpression.scala]: miss copyright

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala [hunk 2]
```diff
@@ -0,0 +1,221 @@
+package org.apache.spark.sql.catalyst.expressions
```
Comment (birdstorm) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala]: ditto

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                      AggregateFunction.FunctionType.First,
+                      ref,
+                      meta.getColumn(ref.getName).getType
+                    )
+              )
+          )
+      case _ =>
+    }
+
+  def transformAggExprToTiAgg(expr: Expression, meta: TiTableInfo, dagRequest: TiDAGRequest): Any =
```
Comment (birdstorm) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala]: have you tried aggregation on unsigned values?

File: core/src/main/scala/org/apache/spark/sql/extensions/rules.scala [hunk 4]
```diff
@@ -30,7 +30,7 @@ case class TiResolutionRule(getOrCreateTiContext: SparkSession => TiContext)(
   sparkSession: SparkSession
 ) extends Rule[LogicalPlan] {
   protected val tiContext: TiContext = getOrCreateTiContext(sparkSession)
-  private lazy val autoLoad = tiContext.autoLoad
+  private lazy val autoLoad = false
```
Comment (birdstorm) [inline core/src/main/scala/org/apache/spark/sql/extensions/rules.scala]: add a comment to this change?

File: core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala [hunk 5]
```diff
@@ -112,7 +112,7 @@ class UnsignedTestSuite extends BaseTiSparkTest {
       List(List(1), List(2), List(3)),
       List(List(0), List(1), List(2), List(3), List(unsignedLongMaxValue))
     )
-    assert(queries.size == answers.size)
+//    assert(queries.size == answers.size)
```
Comment (birdstorm) [inline core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala]: why

File: tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java [hunk 6]
```diff
@@ -34,38 +35,41 @@
   }
 
   public static ArithmeticBinaryExpression plus(Expression left, Expression right) {
-    return new ArithmeticBinaryExpression(PLUS, left, right);
+    return new ArithmeticBinaryExpression(left.dataType, PLUS, left, right);
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java]: I think we should consider type promotion

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java [hunk 7]
```diff
@@ -71,6 +74,10 @@ public TiTableInfo(
     this.charset = charset;
     this.collate = collate;
     this.columns = ImmutableList.copyOf(requireNonNull(columns, "columns is null"));
+    this.columnsMap = new HashMap<>();
+    for (TiColumnInfo col : this.columns) {
+      this.columnsMap.put(col.getName(), col);
```
Comment (birdstorm) [inline tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java]: question: is this map case sensitive?

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala (hunk 3 above)
Comment (zhexuany) [reply core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala]: Yes. Any overflowed values will be prohibit to push down to coprocessor.

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java (hunk 7 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java]: nope. It is case insensitive.

File: tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java (hunk 6 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java]: dataType will be reset in `BasicExpression`.

File: core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala (hunk 5 above)
Comment (zhexuany) [reply core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala]: It is commented for speeding up debug process.

Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala]: if this is commented, the following lines may break.

Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/expression/index/UnsignedTestSuite.scala]: if this is commented, the following lines may break.

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java (hunk 7 above)
Comment (birdstorm) [reply tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java]: then how to make sure we choose the correct column?

File: tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java (hunk 6 above)
Comment (birdstorm) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java]: So any data type is valid here? Where is the logic of reset?

**Comment by birdstorm** [comment]
I recommend that explain details contain more info about types such as decimal and string types.  e.g., `col_decimal@Decimal(20, 3)` @zhexuany

**Comment by zhexuany** [comment]
> I recommend that explain details contain more info about types such as decimal and string types. e.g., `col_decimal@Decimal(20, 3)` @zhexuany

Advice taken.

File: tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java (hunk 7 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/meta/TiTableInfo.java]: You can check `col.getName()`. The `name` of `TiColumnInfo` is always little case. In `getColumn()` of `TiTableInfo`, we always use lower case to get correct `TiColumnInfo`. 
Therefore, we can make sure we always choose the correct column.

**Pull request merged**